### ISSUE TITLE
Spack fails on RHEL7 built in FIPS mode because MD5 isn't FIPS-compliant

### ISSUE THREAD

**Issue by goxberry** [spack/spack#7333]
## Summary

RHEL 7 built in FIPS mode disables insecure hashes like MD5, and this behavior seems to break `spack`.

### Expected Result

Some sort of normal output from `spack --help`.

### Actual Result

```console
[oxberry1@localhost spack]$ spack --help
Traceback (most recent call last):
  File "/home/oxberry1/spack/bin/spack", line 84, in <module>
    import spack.main  # noqa
  File "/home/oxberry1/spack/lib/spack/spack/__init__.py", line 84, in <module>
    import spack.repository
  File "/home/oxberry1/spack/lib/spack/spack/repository.py", line 52, in <module>
    import spack.spec
  File "/home/oxberry1/spack/lib/spack/spack/spec.py", line 122, in <module>
    import spack.store
  File "/home/oxberry1/spack/lib/spack/spack/store.py", line 49, in <module>
    from spack.database import Database
  File "/home/oxberry1/spack/lib/spack/spack/database.py", line 60, in <module>
    from spack.util.crypto import bit_length
  File "/home/oxberry1/spack/lib/spack/spack/util/crypto.py", line 38, in <module>
    size_to_hash = dict((h().digest_size, h) for h in hashes.values())
  File "/home/oxberry1/spack/lib/spack/spack/util/crypto.py", line 38, in <genexpr>
    _size_to_hash = dict((h().digest_size, h) for h in hashes.values())
ValueError: error:060800A3:digital envelope routines:EVP_DigestInit_ex:disabled for fips 
```

### Steps to reproduce the issue

```console
# Build RHEL7 in FIPS mode, clone spack, then
$ spack --help
```

### Information on your system

As far as I can tell, the system is a RHEL7 system built in FIPS mode. OpenSSL is installed, however, running `openssl md5 $FILE` will return an error, and `openssl sha256 $FILE` succeeds and outputs a SHA256 hash. Unfortunately, I don't have that much control over how this system is provisioned due to LLNL policy.

### Conjecture

MD5 isn't FIPS-compliant, so nothing in Spack that uses MD5 is going to work. Unfortunately, AFAICT, many packages in Spack use MD5...

**Comment by tzanio** [comment]
I have the exact same issue on my Linux box at LLNL. @goxberry, were you able to find a work-around?

This is pretty critical for me, as I'm trying to test the spack builds of the CEED distribution packages that we are releasing by the end of the month.

cc: @tgamblin, @acfisher, @v-dobrev

**Comment by v-dobrev** [comment]
Just to confirm: I have the same issue on my RHEL7 LLNL desktop.

**Comment by goxberry** [comment]
@tzanio I don’t see a quick workaround to this issue. Fixing it requires removing or working around any hashing algorithms that aren’t FIPS-compliant. For instance, replacing all of the MD5 hashes with SHA256 equivalents would suffice, but this task would require replacing a thousand or more hashes, which is a lot of files to download and re-hash.

**Comment by scheibelp** [comment]
FWIW if someone can't get to this before I should be able to start looking at it on 3/19

**Comment by scheibelp** [comment]
Does https://github.com/spack/spack/pull/7531 help?

**Comment by tzanio** [comment]
Yes, it does!!

I get some warnings like the one below, but otherwise it works fine:
```
==> Warning: Deprecation warning: md5 is not supported for security use
```

Thanks @scheibelp!

**Comment by v-dobrev** [comment]
#7531 works for me too. Thanks!

**Comment by tgamblin** [comment]
Quick fixed by #7531.  We'll also be replacing MD5 with SHA256 everywhere and completely deprecate MD5 shortly.
